proposals/0030-core-image-story.md

```diff
@@ -0,0 +1,427 @@
+
+## Image Component
+
+### Image located in `src`, in an Astro file
+
+```astro
+---
+import { Image } from 'astro:assets';
+import myImage from "../assets/my_image.png"; // Image is 1600x900
+---
+
+<Image src={myImage} alt="..." />
+```
+
+#### Result
+
+```html
+<!-- Image is optimized, proper attributes are enforced -->
+<img
+  src="/_astro/my_image.hash.webp"
+  width="1600"
+  height="900"
+  decoding="async"
+  loading="lazy"
+  alt="..."
+/>
+```
+
+### Remote images
+
+```astro
+---
+import { Image } from "astro:assets";
+// Image is 1920x1080
+---
+
+<Image src="https://example.com/image.png" alt="..." />
+<!-- ERROR! `width` and `height` are required -->
+
+<Image src="https://example.com/image.png" width={1280} alt="..." />
+<!-- ERROR! `height` is required -->
+
+<Image src="https://example.com/image.png" width={1280} height={720} alt="..." />
+```
+
+#### Result
+
+```html
+<!-- Remote images are not optimized or resized. Nonetheless, you still benefit from the enforced attributes and the guarantee of no CLS, albeit manually. -->
+<img
+  src="https://example.com/image.png"
+  decoding="async"
+  loading="lazy"
+  width="1280"
+  height="720"
+  alt="..."
+/>
+```
+
+## Image in Markdown
+
+```markdown
+![My article cover](./cover.png)
+^ Image is 1280x720
+```
+
+### Result
+
+```html
+<!-- Image is optimized, proper attributes are enforced -->
+<img
+  src="/_astro/cover.hash.webp"
+  width="1280"
+  height="720"
+  decoding="async"
+  loading="lazy"
+  alt="My article cover"
+/>
+```
+
+Remote images are handled as in vanilla Markdown and have no special behavior.
+
+## JavaScript API for using optimized images
+
+```astro
+---
+import { getImage } from "astro:assets";
+import image from "../assets/image.png"; // Image is 1920x1080
+
+const myImage = await getImage({ src: image, width: 1280, height: 720 });
+---
+
+<img src={myImage.src} {...myImage.attributes} />
+```
+
+### Result
+
+```html
+<img
+  src="/_astro/image.hash.webp"
+  width="1280"
+  height="720"
+  loading="lazy"
+  decoding="async"
+/>
+```
+
+## Defining a new image service
+
+### Astro Config
+
+```ts
+import { defineConfig } from "astro/config";
+
+// https://astro.build/config
+export default defineConfig({
+  image: {
+    service: "your-entrypoint", // 'astro/image/services/squoosh' | 'astro/image/services/sharp' | string
+  },
+});
+```
+
+### Local services API
+
+```ts
+import type { LocalImageService } from "astro";
+
+const service: LocalImageService = {
+  getURL(options: ImageTransform) {
+    return `/my_super_endpoint_that_transforms_images?w=${options.width}`;
+  },
+  parseURL(url: URL) {
+    return {
+      width: url.searchParams.get("w"),
+    };
+  },
+  transform(options: ImageTransform) {
+    const { buffer } = mySuperLibraryThatEncodesImages(options);
+
+    return {
+      data: buffer,
+      format: options.format,
+    };
+  },
```

Comment:
Will the compiler throw an error if some functions are not implemented?

Reply:
Sorry, I'm not sure if I understand correctly. If you mean the Astro compiler, no. If you mean `tsc`, yes, as the type require those methods to exists on the object.

Reply:
Yeah sorry for the poor choice of words 😆 

Yes, I meant the TS compiler! Thank you!